apps/mark-scan/fai-100-controller/src/fai_100_controllerd.rs

```diff
@@ -0,0 +1,555 @@
+                    message: format!("Error reading GetFirmwareVersion response: {error}")
+                );
+            }
+        },
+        Err(ref e) if e.kind() == io::ErrorKind::TimedOut => {
+            println!("No data received from IN ENDPOINT");
+        }
+        Err(e) => {
+            log!(
+                event_id: EventId::ControllerHandshakeComplete,
+                message: format!("Error reading echo response: {e:?}"),
+                event_type: EventType::SystemAction,
+                disposition: Disposition::Failure
+            );
+        }
+    }
+
+    Err(io::Error::new(
+        io::ErrorKind::TimedOut,
+        "No echo response received",
+    ))
+}
+
+#[derive(Debug)]
+struct CurrentStatus {
+    button_pressed: ButtonSignal,
+    sip: SipAndPuffSignalStatus,
+    puff: SipAndPuffSignalStatus,
+    sip_puff_device_connected: SipAndPuffDeviceStatus,
+}
+
+struct KeypressSpec {
+    key: keyboard::Key,
+    send_shift: bool,
+}
+
+const fn get_key_for_button_signal(signal: &ButtonSignal) -> Option<KeypressSpec> {
+    let key: keyboard::Key;
+    let mut send_shift: bool = false;
+
+    match signal {
+        ButtonSignal::Select => {
+            key = keyboard::Key::Enter;
+        }
+        ButtonSignal::Left => {
+            key = keyboard::Key::Left;
+        }
+        ButtonSignal::Right => {
+            key = keyboard::Key::Right;
+        }
+        ButtonSignal::Up => {
+            key = keyboard::Key::Up;
+        }
+        ButtonSignal::Down => {
+            key = keyboard::Key::Down;
+        }
+        ButtonSignal::Help => {
+            send_shift = true;
+            key = keyboard::Key::R;
+        }
+        ButtonSignal::RateDown => {
+            key = keyboard::Key::Comma;
+        }
+        ButtonSignal::RateUp => {
+            key = keyboard::Key::Dot;
+        }
+        ButtonSignal::VolumeDown => {
+            key = keyboard::Key::Minus;
+        }
+        ButtonSignal::VolumeUp => {
+            key = keyboard::Key::Equal;
+        }
+        ButtonSignal::Pause | ButtonSignal::Play => {
+            send_shift = true;
+            key = keyboard::Key::P;
+        }
+        ButtonSignal::NoButton => return None,
+    }
+
+    Some(KeypressSpec { key, send_shift })
+}
+
+fn send_keystroke(keypress: &KeypressSpec, keyboard: &mut impl VirtualKeyboard) {
+    if let Err(err) = keyboard.send_keystroke(
+        keypress.key,
+        if keypress.send_shift {
+            &[keyboard::Key::LeftShift]
+        } else {
+            &[]
+        },
+    ) {
+        log!(EventId::UnknownError, "Error sending key: {err}");
+    }
+}
+
+/// Checks new status for changed statuses in button press, sip, puff, and sip & puff device connection.
+/// If changing from inactive to active, sends the appropriate event (keypress or system file update).
+/// If changing from active to inactive or no status change, does nothing.
+fn handle_status_response(
```

Comment:
Similar to `controllerd/handle_command` (https://github.com/votingworks/vxsuite/blob/main/apps/mark-scan/accessible-controller/src/commands.rs#L166), but moved up to the top level because it seemed odd that the command parsing layer also was aware of the `uinput` device and business logic of which controller keys map to which keyboard keys. It also makes boundaries in testing a little more clean. I'm open to considering other options though.